types/validator_set.go

```diff
@@ -42,19 +45,17 @@ type ValidatorSet struct {
 // NewValidatorSet initializes a ValidatorSet by copying over the
 // values from `valz`, a list of Validators. If valz is nil or empty,
 // the new ValidatorSet will have an empty list of Validators.
+// The addresses of vlaidators in `valz` must be unique otherwise the
+// function panics.
 func NewValidatorSet(valz []*Validator) *ValidatorSet {
-	validators := make([]*Validator, len(valz))
-	for i, val := range valz {
-		validators[i] = val.Copy()
-	}
-	sort.Sort(ValidatorsByAddress(validators))
-	vals := &ValidatorSet{
-		Validators: validators,
+	vals := &ValidatorSet{}
+	err := vals.UpdateWithChangeSet(valz)
```

Comment:
It's interesting that this doesn't change the behaviour now. IIUC, calling `NewValidatorSet` would initialize priorities with one call of `incrementProposerPriority` (basically setting all prios to the voting power of that validator and decrement the mostest). Now, we detect duplicates and additionally set all initial prios to -1.125*total (then rescale, shift by average via `UpdateWithChangeSet`) and then do the `incrementProposerPriority` (which rescales / shifts again).
The test suggest this doesn't have any side effects. And as we are not constantly calling `NewValidatorSet`, I think it is negligable to do the rescaling/shifting several times here.

Reply:
If we do not want that, and to catch the duplicates, we could call the private helper functions (processChanges, verifyUpdates) here directly (would duplicate some of the code of `UpdateWithChangeSet`)

Reply:
> It's interesting that this doesn't change the behaviour now. IIUC, calling `NewValidatorSet` would initialize priorities with one call of `incrementProposerPriority` (basically setting all prios to the voting power of that validator and decrement the mostest). Now, we detect duplicates and additionally set all initial prios to -1.125*total (then rescale, shift by average via `UpdateWithChangeSet`) and then do the `incrementProposerPriority` (which rescales / shifts again).

Yes true it's an overhead. But I think we will remove the rescale from incrementProposerPriority() once we know the priority window bound. 

> The tests suggest this doesn't have any side effects. And as we are not constantly calling `NewValidatorSet`, I think it is negligable to do the rescaling/shifting several times here.

Yes, this is what creates the validator set so only called once. After this we have the updates that do pretty much the same.

>If we do not want that, and to catch the duplicates, we could call the private helper functions (processChanges, verifyUpdates) here directly (would duplicate some of the code of `UpdateWithChangeSet`)

My concern is not really the extra rescaling/ centering. That I think will get better as I mentioned above. The problem right now is the different semantics of the parameter: - --
- NewValidatorSet() : add all validators in list, vs.
- UpdateWithChangeSet(): if validator exist either update or delete, if not add

So right now, when we call NewValidatorSet() with a validator with power 0 we fail (as we should) but the error says something like "cannot delete validator with address blah.."
I will fix this soon.

Reply:
Made some changes so we have better errors.